relayer/chains/substrate/event_parser.go

```diff
@@ -0,0 +1,221 @@
+	"go.uber.org/zap/zapcore"
+)
+
+// ibcMessage is the type used for parsing all possible properties of IBC messages
+type ibcMessage struct {
+	eventType string
+	info      ibcMessageInfo
+}
+
+type ibcMessageInfo interface {
+	MarshalLogObject(enc zapcore.ObjectEncoder) error
+}
+
+// ibcMessagesFromTransaction parses all events within a transaction to find IBC messages
+func (scp *SubstrateChainProcessor) handleIBCMessagesFromEvents(ibcEvents rpcclienttypes.IBCEventsQueryResult, height uint64, c processor.IBCMessagesCache) error {
+	for i := 0; i < len(ibcEvents); i++ {
+		for eType, data := range ibcEvents[i] {
+			var info ibcMessageInfo
+			var eventType string
+
+			marshalled, _ := json.Marshal(data)
+
+			switch eType {
+
+			case "CreateClient", "UpgradeClient", "ClientMisbehaviour":
+				ce := clientInfo{}
+				json.Unmarshal(marshalled, &ce)
+				info = &ce
+				eventType = eType
+
+			case "UpdateClient":
+				ce := clientUpdateInfo{}
+				json.Unmarshal(marshalled, &ce)
+				info = &ce
+				eventType = eType
+
+			case "OpenInitConnection", "OpenTryConnection", "OpenAckConnection", "OpenConfirmConnection":
+				ce := connectionEvent{}
+				json.Unmarshal(marshalled, &ce)
+				info = &connectionInfo{
+					Height:               height,
+					ConnID:               ce.connectionID,
+					ClientID:             ce.clientID,
+					CounterpartyConnID:   ce.counterpartyConnectionID,
+					CounterpartyClientID: ce.counterpartyClientID,
+				}
+				eventType = eType
+
+			case "OpenInitChannel", "OpenTryChannel", "OpenAckChannel", "OpenConfirmChannel", "CloseInitChannel", "CloseConfirmChannel":
+				ce := channelEvent{}
+				json.Unmarshal(marshalled, &ce)
+				info = &channelInfo{
+					Height:                height,
+					PortID:                ce.portID,
+					ChannelID:             strconv.FormatUint(ce.channelID, 10),
+					CounterpartyPortID:    strconv.FormatUint(ce.counterpartyPortID, 10),
+					CounterpartyChannelID: ce.counterpartyChannelID,
+					ConnID:                ce.connectionID,
+					CounterpartyConnID:    ce.counterpartyChannelID,
+					// TODO: how to populate order
+					// Order: ,
+					// TODO: how to populate version
+					// Version: ,
+				}
+				eventType = eType
+
+			case "SendPacket", "ReceivePacket", "WriteAcknowledgement", "AcknowledgePacket", "TimeoutPacket", "TimeoutOnClosePacket":
+				ce := packetEvent{}
+				json.Unmarshal(marshalled, &ce)
+				info = &packetInfo{
+					Height:        height,
+					Sequence:      ce.sequence,
+					SourcePort:    ce.sourcePort,
+					SourceChannel: strconv.FormatUint(ce.sourceCannel, 10),
+					DestPort:      ce.destinationPort,
+					DestChannel:   strconv.FormatUint(ce.destinationChannel, 10),
+					// TODO: how to populate channel order
+					// ChannelOrder: ,
+					Data:             ce.data,
+					TimeoutHeight:    ce.timeoutHeight,
+					TimeoutTimestamp: ce.timeoutTimestamp,
+					// TODO: how to populate Ack
+					// Ack: ,
+
+				}
+				eventType = eType
```

Comment:
The `PathProcessor` depends on the eventType being in the format of `ibc-go`. For example, instead of `SendPacket`, it should be `chantypes.EventTypeSendPacket` (which is `send_packet`). A helper might be nice to translate from the substrate ibc event type to the ibc-go event type.

Reply:
Here's a consolidated place to find them in the cosmos event parser https://github.com/cosmos/relayer/blob/91182c9c766bfe3f2825255170a094dfd7bad287/relayer/chains/cosmos/event_parser.go#L86